### PR TITLE
Fail on unknown JDBC type

### PR DESCRIPTION
**Pull request by kokosing**
Fail on unknown JDBC type

### CODE REVIEW

File: presto-tests/src/main/java/io/prestosql/tests/sql/TestTable.java [hunk 1]
```diff
@@ -13,21 +13,48 @@
...
+        }
+        try {
+            sqlExecutor.execute(format("INSERT INTO %s VALUES %s", name, data.get()));
+        }
+        catch (Exception e) {
+            try {
+                close();
+            }
+            catch (Exception innerException) {
+                if (e != innerException) {
```
Comment (sopel39) [inline presto-tests/src/main/java/io/prestosql/tests/sql/TestTable.java]: Could it really be the same exception instance?

Comment (kokosing) [reply presto-tests/src/main/java/io/prestosql/tests/sql/TestTable.java]: Yes, I know. I think I can call it a `Presto convention`.

File: presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/BaseJdbcClient.java [hunk 2]
```diff
@@ -248,12 +249,22 @@ public void destroy()
...
                     if (columnMapping.isPresent()) {
-                        String columnName = resultSet.getString("COLUMN_NAME");
                         boolean nullable = (resultSet.getInt("NULLABLE") != columnNoNulls);
                         columns.add(new JdbcColumnHandle(columnName, typeHandle, columnMapping.get().getType(), nullable));
                     }
+                    else {
+                        UnsupportedTypeHandlingStrategy unsupportedTypeHandlingStrategy = getUnsupportedTypeHandlingStrategy(session);
+                        switch (unsupportedTypeHandlingStrategy) {
+                            case IGNORE:
+                                break;
```
Comment (findepi) [inline presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/BaseJdbcClient.java]: `continue` (please add a test)

File: presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/BaseJdbcClient.java [hunk 3]
```diff
@@ -248,12 +249,22 @@ public void destroy()
...
+                            case FAIL:
+                                throw new PrestoException(JDBC_ERROR, "Unsupported data type for column: " + columnName);
```
Comment (findepi) [inline presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/BaseJdbcClient.java]: incl `typeHandle`

File: presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/JdbcConnector.java [hunk 4]
```diff
@@ -142,6 +148,14 @@ public ConnectorAccessControl getAccessControl()
         return procedures;
     }
 
+    @Override
+    public List<PropertyMetadata<?>> getSessionProperties()
+    {
+        ImmutableList.Builder builder = ImmutableList.<PropertyMetadata<?>>builder();
```
Comment (findepi) [inline presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/JdbcConnector.java]: raw usage of a generic class

File: presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/BaseJdbcConfig.java [hunk 5]
```diff
@@ -124,4 +127,18 @@ public BaseJdbcConfig setCaseInsensitiveNameMatchingCacheTtl(Duration caseInsens
...
         return this;
     }
+
+    @NotNull
+    public UnsupportedTypeHandlingStrategy getUnsupportedTypeHandlingStrategy()
+    {
+        return unsupportedTypeHandlingStrategy;
+    }
+
+    @Config("unsupported-type.handling-strategy")
```
Comment (findepi) [inline presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/BaseJdbcConfig.java]: `unsupported-type.handling-strategy=FAIL` is not nearly as clear as `fail-on-unsupported-types=true`.

I understand the enum is a preparation for an option to map unsupported columns to varchar (see also https://github.com/prestosql/presto/pull/186)
I think the option name needs to be revisited.

File: presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/BaseJdbcConfig.java [hunk 6]
```diff
@@ -32,6 +34,7 @@
     private String passwordCredentialName;
     private boolean caseInsensitiveNameMatching;
     private Duration caseInsensitiveNameMatchingCacheTtl = new Duration(1, MINUTES);
+    private UnsupportedTypeHandlingStrategy unsupportedTypeHandlingStrategy = FAIL;
```
Comment (findepi) [inline presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/BaseJdbcConfig.java]: Do we have already enough data points to determine what the default behavior should actually be?

File: presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/BaseJdbcClient.java [hunk 7]
```diff
@@ -63,6 +63,7 @@
 import static com.google.common.collect.ImmutableMap.toImmutableMap;
```
Comment (kokosing) [inline presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/BaseJdbcClient.java]: query should fail only when user actually tries to read data from such column

**Comment by findepi** [comment]
I propose a different new behavior:
- expose all columns, including those unsupported (perhaps as `unknown` type)
- fail query only when it accesses unsupported columns (either explicitly, or `SELECT *`)

Regarding default behavior, there are different use-cases for `SELECT *`:

1. copying/migrating data (usually `CREAT TABLE .. AS SELECT * FROM ..`)
   - in this case, the new behavior is more appropriate
2. poking around (usually `SELECT * FROM .. LIMIT 1`)
   - in this case, the current behavior is more appropriate

Since current behavior is more appropriate in certain use-cases, I would *not* change current default behavior.

**Comment by kokosing** [comment]
What about conversion `unknown` types to varchar? I think we could add only a session property that would be always set to false (no config property to overwrite that). So user could peek the data, knowing the fact that column data type is unsupported and it may change in future.

@findepi What do you think?

**Comment by electrum** [comment]
Seems reasonable to me.

**Comment by kokosing** [comment]
Replaced with #1182

**Pull request closed without merging**